README.rst
Django MSSQL Database Backend
=============================

`django_pytds`_ provies a Django database backend for Microsoft SQL Server.

It is based on django-mssql but replaces the ADO component with the PyTDS component. No PyODBC of FreeTDS required.

Documentation is available at `django-mssql.readthedocs.org`_.

Requirements
------------

    * Python 2.7
    * PyTDS  (https://github.com/denisenkom/pytds)

Not required but good to have as PyTDS claims that it makes it faster:
    * bitarray


SQL Server Versions
-------------------

Supported Versions:
    * 2008
    * 2008r2
    * 2012

The SQL Server version will be detected upon initial connection.

Django Version
--------------

I haven't tested much. This is still work in progress but it seems to work pretty fine so far. The original project django-mssql works on the following:

	* Django 1.6
	* Django 1.7 (schema migrations may not be fully supported)


django-mssql 1.4 supports Django 1.4 and 1.5. 
(so I suppose we do, too.)


Django Setup
------------
In Django Settings, use::

	'default': {
		'ENGINE': 'django_pytds',
		'NAME': 'DBNAME',
		'HOST': 'DBHOST',
		'USER': 'USERNAME',
		'PASSWORD': 'PASSWORD',
		'OPTIONS': {
			'autocommit': True/False   (Default: False)
			'use_mars':  True/False    (Default: False)
			'readonly':  True/False    (Default: False)
		},
	}

The options part is optional.

Notes
-----

I honestly don't have time to maintain this project. If you feel like you want to help, please go ahead. This was just implemented to get a project moving and it fulfils the needs perfectly at this point, so I don't think I'll be adding more to it. It's very easy to fix issues on this project than to setup PyODBC and FreeTDS and PyTDS is alive and gets updated regularly. Patch/Fix this instead of wasting time on the ODBC driver.

References
----------

    * PyTDS on github: https://github.com/denisenkom/pytds
    * Django-mssql on PyPi: http://pypi.python.org/pypi/django-mssql
    * DB-API 2.0 specification: http://www.python.org/dev/peps/pep-0249/

.. _`PyTDS`: https://github.com/denisenkom/pytds
.. _`Django-mssql`: https://bitbucket.org/Manfre/django-mssql
.. _django-mssql.readthedocs.org: http://django-mssql.readthedocs.org/
.. _PyWin32: http://sourceforge.net/projects/pywin32/


docs/changelog.txt
Changelog
=========

v1.7 (In Development)
---------------------

- Enabled bulk_insert database feature. :issue:`6` (Thanks João Pedro Francese)
- Fixed string representation for Decimals in format '2.82E+3'. :issue:`55`
  (Thanks João Pedro Francese)

v1.6.1
------

- Fixed issue with setup.py that prevented installing on Python 3.4. :issue:`60`


v1.6
----

- SQL Server 2005 is no longer supported. It should not be used.
- Added support for Django 1.7.

  - Schema migrations is a new Django feature that may have unexpected issues
    that could result in data loss or destruction of the database schema. You
    should inspect all generated SQL before manually applying to your production
    database.

- Password is masked from the connection string if there are connection errors.
  Thanks Martijn Pieters.
- Added :ref:`project-goals` and documentation geared toward those wanting to
  contribute to Django-mssql. See :ref:`contributing`.
- Removed the `dbgui` management command, which didn't work anyway.

v1.5.1
------

- Datetime strings are now in a format that should work regardless of SQL language/format setting. :issue:`57`
- The default value for :setting:`use_legacy_date_fields` has changed to 'False'. This setting will be removed in a
  later version. Once removed, any usage of the legacy 'datetime' datatype will require using the provided
  'Legacy*Field' model fields.
- SQL Server 2005 has not been tested in a while and support will officially be removed in the next release.
- Decimals are passed through ADO as strings to avoid rounding to four places. :issue:`55`.
- Database introspection will now identify a char type with a length over 8000 as a ``TextField``. :issue:`53`
- Minor changes to make it possible to subclass django-mssql for use on a non-Windows platform.

v1.5
----

- This version only supports Django v1.6. Use a previous version if you are using Django v1.5 or earlier.
- Added ``BinaryField`` as ``varbinary(max)``
- Refactored DatabaseOperations.sql_flush to use a faster method of disabling/enabling constraints.
- ``sqlserver_ado.fields.DateTimeField`` will now do a better job of returning an aware or naive datetime.
- DateTime truncation will now support any ``datetime2`` value without the potential of a datediff overflow exception.
- Many improvements to regex_clr

  - Updated regex_clr to Visual Studio 2010
  - Patterns and fields with a length greater than 4000 will now work.
  - If pattern or input are NULL, it will not match instead of raising an exception.
  - regex_clr auto-installs during test database creation
  - Added command :ref:`command_install_regex_clr` to simplify enabling regex support for any database.

- Database introspection will now identify a ``varchar(max)`` field as a ``TextField``.
- DatabaseIntrospection.get_indexes now properly finds regular, non-unique, non-primary indices.
- Complete refactor of django-mssql test suite and abandoned tests that are covered by Django test suite.
- Test database is dropped in a more forcible way to avoid "database in use" errors.

v1.4
----

- Support for Django v1.3 has been removed.
- Corrected DB-API 2 testing documentation.
- Fixed issue with slicing logic that could prevent the compiler from finding
  and mapping column aliases properly.
- Improved the "return ID from insert" logic so it can properly extract the
  column data type from user defined fields with custom data type strings.
- Fixed case for identifiers in introspection. Thanks Mikhail Denisenko.
- Added option :setting:`use_legacy_date_fields` (defaults to True) to allow changing the DatabaseCreation.data_types to
  not use the Microsoft preferred date data types that were added with SQL Server 2008. :issue:`31`
- Improved accuracy of field type guessing with inspectdb. See :ref:`introspecting-custom-fields`
- Fixed issue with identity insert using a cursor to the wrong database in a multi-database environment. Thanks Mikhail 
  Denisenko
- Fixed constraint checking. :issue:`35` Thanks Mikhail Denisenko
- Enabled ``can_introspect_autofield`` database feature. :djangoticket:`21097`
- Any date related field should now return from the database as the appropriate Python type, instead of always being a
  datetime.
- Backend now supports doing date lookups using a string. E.g. ``Article.objects.filter(pub_date__startswith='2005')``
- ``check_constraints`` will now check all disabled and enabled constraints. This change was made to match the behavior
  tested by ``backends.FkConstraintsTests.test.test_check_constraints``.
- Improved ``date_interval_sql`` support for the various date/time related datatypes. The ``timedelta`` value will
  control whether the database will ``DATEADD`` using ``DAY`` or ``SECOND``. Trying to add seconds to a ``date``, or
  days to a ``time`` will generate database exceptions.
- Fixed issue with provider detection that prevented ``DataTypeCompatibility=80`` from being automatically added to the 
  connection string for the native client providers.
- Fixed SQL generation error that occured when ordering the query based upon a column that is not being returned.
- Added savepoint support. MS SQL Server doesn't support savepoint commits and will no-op it. Other databases, e.g.
  postgresql, mostly use it as a way of freeing server resources in the middle of a transaction. Thanks Martijn Pieters.
- Minor cleanup of limit/offset SQL mangling to allow custom aggregates that require multiple column replacements.
  :issue:`40` Thanks Martijn Pieters for initial patch and tests.
- Savepoints cannot be used with MARS connections. :issue:`41`

v1.3.1
------

- Ensure Django knows to re-enable constraints. :issue:`29`

v1.3
----

- Backend now supports returning the ID from an insert without needing an additional query. This is disabled
  for SQL Server 2000 (assuming that version still works with this backend). :issue:`17`

  - This will work even if the table has a trigger. :issue:`20`


django_pytds/base.py
"""Microsoft SQL Server database backend for Django."""
from __future__ import absolute_import, unicode_literals

import warnings

from django.core.exceptions import ImproperlyConfigured, ValidationError
from django.db.backends import BaseDatabaseWrapper, BaseDatabaseFeatures, BaseDatabaseValidation, BaseDatabaseClient
from django.db.utils import IntegrityError as DjangoIntegrityError, \
    InterfaceError as DjangoInterfaceError
from django.utils.functional import cached_property
from django.utils import six

from . import pytdsdbapi as Database        # PyTDS stuff

from .introspection import DatabaseIntrospection
from .creation import DatabaseCreation
from .operations import DatabaseOperations
from .schema import DatabaseSchemaEditor

try:
    import pytz
except ImportError:
    pytz = None

IntegrityError = Database.IntegrityError


class DatabaseFeatures(BaseDatabaseFeatures):
    uses_custom_query_class = True
    has_bulk_insert = False

    # DateTimeField doesn't support timezones, only DateTimeOffsetField
    supports_timezones = False
    supports_sequence_reset = False

    can_return_id_from_insert = True

    supports_regex_backreferencing = False

    supports_tablespaces = True

    # Django < 1.7
    ignores_nulls_in_unique_constraints = False
    # Django >= 1.7
    supports_nullable_unique_constraints = False
    supports_partially_nullable_unique_constraints = False

    can_introspect_autofield = True
    can_introspect_small_integer_field = True

    supports_subqueries_in_group_by = False

    allow_sliced_subqueries = False

    uses_savepoints = True

    supports_paramstyle_pyformat = False

    closed_cursor_error_class = DjangoInterfaceError

    # connection_persists_old_columns = True

    requires_literal_defaults = True

    @cached_property
    def has_zoneinfo_database(self):
        return pytz is not None

    # Dict of test import path and list of versions on which it fails
    failing_tests = {
        # Some tests are known to fail with django-mssql.
        'aggregation.tests.BaseAggregateTestCase.test_dates_with_aggregation': [(1, 6), (1, 7)],
        'aggregation_regress.tests.AggregationTests.test_more_more_more': [(1, 6), (1, 7)],

        # MSSQL throws an arithmetic overflow error.
        'expressions_regress.tests.ExpressionOperatorTests.test_righthand_power': [(1, 7)],

        # The migrations and schema tests also fail massively at this time.
        'migrations.test_operations.OperationTests.test_alter_field_pk': [(1, 7)],

    }


def is_ip_address(value):
    """
    Returns True if value is a valid IP address, otherwise False.
    """
    # IPv6 added with Django 1.4
    from django.core.validators import validate_ipv46_address as ip_validator

    try:
        ip_validator(value)
    except ValidationError:
        return False
    return True


def connection_string_from_settings():
    from django.conf import settings
    db_settings = getattr(settings, 'DATABASES', {}).get('default', None)
    return make_connection_string(db_settings)


def make_connection_string(settings):
    db_name = settings['NAME'].strip()
    db_host = settings['HOST'] or '127.0.0.1'
    db_port = settings['PORT']
    db_user = settings['USER']
    db_password = settings['PASSWORD']
    options = settings.get('OPTIONS', {})

    if len(db_name) == 0:
        raise ImproperlyConfigured("You need to specify a DATABASE NAME in your Django settings file.")

    # Connection strings courtesy of:
    # http://www.connectionstrings.com/?carrier=sqlserver

    # If a port is given, force a TCP/IP connection. The host should be an IP address in this case.
    if db_port:
        if not is_ip_address(db_host):
            raise ImproperlyConfigured("When using DATABASE PORT, DATABASE HOST must be an IP address.")
        try:
            db_port = int(db_port)
        except ValueError:
            raise ImproperlyConfigured("DATABASE PORT must be a number.")
        db_host = '{0},{1};Network Library=DBMSSOCN'.format(db_host, db_port)

    # If no user is specified, use integrated security.
    if db_user != '':
        auth_string = 'UID={0};PWD={1}'.format(db_user, db_password)
    else:
        auth_string = 'Integrated Security=SSPI'

    parts = [
        'DATA SOURCE={0};Initial Catalog={1}'.format(db_host, db_name),
        auth_string
    ]

    if not options.get('provider', None):
        options['provider'] = 'sqlncli10'

    parts.append('PROVIDER={0}'.format(options['provider']))

    extra_params = options.get('extra_params', '')

    if 'sqlncli' in options['provider'].lower() and 'datatypecompatibility=' not in extra_params.lower():
        # native client needs a compatibility mode that behaves like OLEDB
        parts.append('DataTypeCompatibility=80')

    if options.get('use_mars', True) and 'mars connection=' not in extra_params.lower():
        parts.append('MARS Connection=True')

    if extra_params:
        parts.append(options['extra_params'])

    return ";".join(parts)


VERSION_SQL2000 = 8
VERSION_SQL2005 = 9
VERSION_SQL2008 = 10
VERSION_SQL2012 = 11


class DatabaseWrapper(BaseDatabaseWrapper):
    vendor = 'microsoft'

    Database = Database

    operators = {
        "exact": "= %s",
        "iexact": "LIKE %s ESCAPE '\\'",
        "contains": "LIKE %s ESCAPE '\\'",
        "icontains": "LIKE %s ESCAPE '\\'",
        "gt": "> %s",
        "gte": ">= %s",
        "lt": "< %s",
        "lte": "<= %s",
        "startswith": "LIKE %s ESCAPE '\\'",
        "endswith": "LIKE %s ESCAPE '\\'",
        "istartswith": "LIKE %s ESCAPE '\\'",
        "iendswith": "LIKE %s ESCAPE '\\'",
    }

    def __init__(self, *args, **kwargs):
        self.use_transactions = kwargs.pop('use_transactions', None)

        super(DatabaseWrapper, self).__init__(*args, **kwargs)

        try:
            self.command_timeout = int(self.settings_dict.get('COMMAND_TIMEOUT', 30))
        except ValueError:
            self.command_timeout = 30

        options = self.settings_dict.get('OPTIONS', {})
        try:
            self.cast_avg_to_float = not bool(options.get('disable_avg_cast', False))
        except ValueError:
            self.cast_avg_to_float = False

        USE_LEGACY_DATE_FIELDS_DEFAULT = False
        try:
            self.use_legacy_date_fields = bool(options.get('use_legacy_date_fields', USE_LEGACY_DATE_FIELDS_DEFAULT))
        except ValueError:
            self.use_legacy_date_fields = USE_LEGACY_DATE_FIELDS_DEFAULT

        if self.use_legacy_date_fields:
                warnings.warn(
                    "The `use_legacy_date_fields` setting has been deprecated. "
                    "The default option value has changed to 'False'. "
                    "If you need to use the legacy SQL 'datetime' datatype, "
                    "you must replace them with the provide model field.",
                    DeprecationWarning)

        self.features = DatabaseFeatures(self)
        self.ops = DatabaseOperations(self)
        self.client = BaseDatabaseClient(self)
        self.creation = DatabaseCreation(self)
        self.introspection = DatabaseIntrospection(self)
        self.validation = BaseDatabaseValidation(self)

    def get_connection_params(self):
        """Returns a dict of parameters suitable for get_new_connection."""
        settings_dict = self.settings_dict.copy()
        if settings_dict['NAME'] == '':
            from django.core.exceptions import ImproperlyConfigured
            raise ImproperlyConfigured(
                "settings.DATABASES is improperly configured. "
                "Please supply the NAME value.")
        if not settings_dict['NAME']:
            # if _nodb_connection, connect to master
            settings_dict['NAME'] = 'master'

        autocommit = settings_dict.get('OPTIONS', {}).get('autocommit', False)
        return {
            'connection_string': make_connection_string(settings_dict),
            'timeout': self.command_timeout,
            'use_transactions': not autocommit,
        }

    def get_new_connection(self, conn_params):
        """Opens a connection to the database."""
        self.__connection_string = conn_params.get('connection_string', '')
        settings_dict = self.settings_dict.copy()
        db_name = settings_dict['NAME'].strip()
        db_host = settings_dict['HOST'] or '127.0.0.1'
        db_port = settings_dict['PORT']
        db_user = settings_dict['USER']
        db_password = settings_dict['PASSWORD']
        autocommit = settings_dict.get('OPTIONS', {}).get('autocommit', False)
        use_mars = settings_dict.get('OPTIONS', {}).get('use_mars', False)
        readonly = settings_dict.get('OPTIONS', {}).get('readonly', False)

        conn = Database.connect(server=db_host, database=db_name, port=db_port, user=db_user, password=db_password,
            timeout=self.command_timeout, use_transactions=not autocommit, use_mars=use_mars, readonly=readonly)

        return conn

    def init_connection_state(self):
        """Initializes the database connection settings."""
        # if 'mars connection=true' in self.__connection_string.lower():
        #     # Issue #41 - Cannot use MARS with savepoints
        #     self.features.uses_savepoints = False
        # cache the properties on the connection
        self.features.can_return_id_from_insert = True
        return
        
        self.connection.adoConnProperties = dict([(x.Name, x.Value) for x in self.connection.adoConn.Properties])

        unsupported_sql = False
        if self.is_sql2000(make_connection=False):
            # SQL 2000 doesn't support the OUTPUT clause
            self.features.can_return_id_from_insert = False
            unsupported_sql = True
        elif self.is_sql2005(make_connection=False):
            unsupported_sql = True

        if unsupported_sql:
            warnings.warn(
                "This version of MS SQL server is no longer tested with "
                "django-mssql and not officially supported/maintained.",
                DeprecationWarning)

    def create_cursor(self):
        """Creates a cursor. Assumes that a connection is established."""
        cursor = self.connection.cursor()
        return cursor

    def _set_autocommit(self, value):
        self.connection.autocommit = value          # PyTDS stuff
        #self.connection.set_autocommit(value)

    def __get_dbms_version(self, make_connection=True):
        """
        Returns the 'DBMS Version' string, or ''. If a connection to the database has not already
        been established, a connection will be made when `make_connection` is True.
        """
        if not self.connection and make_connection:
            self.connect()
        return self.connection.adoConnProperties.get('DBMS Version', '') if self.connection else ''

    def is_sql2000(self, make_connection=True):
        """
        Returns True if the current connection is SQL2000. Establishes a
        connection if needed when make_connection is True.
        """
        return self.__get_dbms_version(make_connection).startswith(six.text_type(VERSION_SQL2000))

    def is_sql2005(self, make_connection=True):
        """
        Returns True if the current connection is SQL2005. Establishes a
        connection if needed when make_connection is True.
        """
        return self.__get_dbms_version(make_connection).startswith(six.text_type(VERSION_SQL2005))

    def is_sql2008(self, make_connection=True):
        """
        Returns True if the current connection is SQL2008. Establishes a
        connection if needed when make_connection is True.
        """
        return self.__get_dbms_version(make_connection).startswith(six.text_type(VERSION_SQL2008))

    def disable_constraint_checking(self):
        """
        Turn off constraint checking for every table
        """
        if self.connection:
            cursor = self.connection.cursor()
        else:
            cursor = self._cursor()
        cursor.execute('EXEC sp_MSforeachtable "ALTER TABLE ? NOCHECK CONSTRAINT all"')
        return True

    def enable_constraint_checking(self):
        """
        Turn on constraint checking for every table
        """
        if self.connection:
            cursor = self.connection.cursor()
        else:
            cursor = self._cursor()
        # don't check the data, just turn them on
        cursor.execute('EXEC sp_MSforeachtable "ALTER TABLE ? WITH NOCHECK CHECK CONSTRAINT all"')

    def check_constraints(self, table_names=None):
        """
        Check the table constraints.
        """
        if self.connection:
            cursor = self.connection.cursor()
        else:
            cursor = self._cursor()
        if not table_names:
            cursor.execute('DBCC CHECKCONSTRAINTS WITH ALL_CONSTRAINTS')
            if cursor.description:
                raise DjangoIntegrityError(cursor.fetchall())
        else:
            qn = self.ops.quote_name
            for name in table_names:
                cursor.execute('DBCC CHECKCONSTRAINTS({0}) WITH ALL_CONSTRAINTS'.format(
                    qn(name)
                ))
                if cursor.description:
                    raise DjangoIntegrityError(cursor.fetchall())

    # # MS SQL Server doesn't support explicit savepoint commits; savepoints are
    # # implicitly committed with the transaction.
    # # Ignore them.
    def _savepoint_commit(self, sid):
        try:
            queries_log = self.queries_log   # Django 1.8+
        except AttributeError:
            queries_log = self.queries       # Django <1.8
        if queries_log:
            queries_log.append({
                'sql': '-- RELEASE SAVEPOINT %s -- (because assertNumQueries)' % self.ops.quote_name(sid),
                'time': '0.000',
            })

    def is_usable(self):
        try:
            # Use a mssql cursor directly, bypassing Django's utilities.
            with self.connection.cursor() as cursor:
                cursor.execute("SELECT 1")
        except self.Database.Error:
            return False
        else:
            return True

    def schema_editor(self, *args, **kwargs):
        return DatabaseSchemaEditor(self, *args, **kwargs)
